sound/soc/sof/pcm.c

```diff
@@ -318,6 +318,19 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 	snd_sof_pcm_platform_trigger(sdev, substream, cmd);
 
 	/* send IPC to the DSP */
+	/*
+	 * The trigger action is in non-atomic environment, which
+	 * is set in topology parsing. Here is the details.
+	 * In soc_new_pcm(), as only FE PCM will really create pcm
+	 * (BE no_pcm is 1), so the pcm->nonatomic will be set to the
+	 * FE link's nonatomic in ASoC:
+	 * pcm->nonatomic = rtd->dai_link->nonatomic
+	 * when playback/capture, function __snd_pcm_stream_lock_mode()
+	 * will check whether pcm nonatomic is set or not. If the pcm is
+	 * set to nonaotmic, alsa will set the environment as nonatomic.
+	 * So as SOF FE has already set it as nonatomic in the topology,
+	 * calling sof_ipc_tx_message() is safe.
+	 */
```

Comment:
Sorry @libinyang, I find your answers convoluted. I was expecting something like

The FE can be triggered in a non-atomic environment because ..... The non-atomic property is set in the DAI link and copied into the PCM fields.

The BE link does not set the non-atomic property because ...

Reply:
Got it. I will change the comments.